netket/experimental/vqs/MinSR.py

```diff
@@ -0,0 +1,226 @@
+    chunk_size: int,
+    grad_chunk_size: int,
+    r_cond: float = 1e-12,
+) -> Tuple[Stats, PyTree]:
+
+    """Calculates the expectation value of an operator and the gradient update
+    according to the MinSR algorithm in `Chen et al. <https://arxiv.org/pdf/2302.01941.pdf>`
+
+    Args:
+        vstate: the variational state
+        Ô : a hermitian operator
+        chunk_size : An integer over which the forward pass is chunked
+        grad_chunk_size : An integer over which the forward and backward pass are chunked
+        r_cond: The lowest value eigenvalues of the NTK that are kept
+
+    Returns:
+        A tuple containing the statistics of Ô and the gradient update according to the MinSR algorithm
+    """
+
+    σ, args = get_local_kernel_arguments(vstate, Ô)
+
+    local_estimator_fun = get_local_kernel(vstate, Ô, chunk_size)
+
+    if jnp.ndim(σ) != 2:
+        σ = σ.reshape((-1, σ.shape[-1]))
+
+    Ō, elocs, grad_mean = expect_MinSR(
+        chunk_size,
+        jacobian_chunk_size,
+        local_estimator_fun,
+        vstate._apply_fun,
+        vstate.parameters,
+        vstate.model_state,
+        σ,
+        args,
+    )
+
+    NTK = compute_NTK(
+        jacobian_chunk_size,
+        vstate._apply_fun,
+        vstate.parameters,
+        vstate.model_state,
+        σ,
+        grad_mean,
+    )
+
+    Ō_grad = grad_MinSR(
+        jacobian_chunk_size,
+        r_cond,
+        vstate._apply_fun,
+        elocs,
+        NTK,
+        vstate.parameters,
+        vstate.model_state,
+        σ,
+        grad_mean,
+    )
+
+    return Ō, Ō_grad
+
+
+@partial(jax.jit, static_argnums=(0, 1, 2, 3))
+def expect_MinSR(
+    chunk_size: int,
+    jcs: int,
+    local_value_kernel_chunked: Callable,
+    model_apply_fun: Callable,
+    parameters: PyTree,
+    model_state: PyTree,
+    σ: jnp.ndarray,
+    local_value_args: PyTree,
+) -> Tuple[PyTree, PyTree]:
+
+    n_samples = σ.shape[0] * mpi.n_nodes
+
+    O_loc = local_value_kernel_chunked(
+        model_apply_fun,
+        {"params": parameters, **model_state},
+        σ,
+        local_value_args,
+        chunk_size=chunk_size,
+    )
+
+    Ō = statistics(O_loc.reshape(σ.shape[:-1]).T)
+
+    O_loc -= Ō.mean
+
+    def grad(w, σ):
+
+        return model_apply_fun({"params": w, **model_state}, σ) / n_samples
+
+    vjp_fun_chunked = nkjax.vjp_chunked(
+        grad,
+        parameters,
+        σ,
+        conjugate=False,
+        chunk_size=jcs,
+        chunk_argnums=1,
+        nondiff_argnums=1,
+    )
+
+    grad_mean = vjp_fun_chunked(
+        jnp.ones_like(O_loc),
+    )[0]
+
+    return Ō, O_loc / n_samples, grad_mean
+
+
+def compute_NTK(
+    jcs: int,
+    model_apply_fun: Callable,
+    parameters: PyTree,
+    model_state: PyTree,
+    σ: jnp.ndarray,
+    grad_mean: jnp.ndarray,
+) -> Tuple[PyTree, PyTree]:
+
+    n_samples = len(σ)
+
+    NTK = jnp.zeros([n_samples, n_samples], dtype="complex")
+
+    for i in range(n_samples // jcs):
```

Comment:
isn't this double for loop very inefficient?

Reply:
Yes it's most efficient to compute the NTK all in one go but you need to chunk the matrix into parts if the gradient function takes too much memory.